src/rpc/protocol.cpp

```diff
@@ -37,7 +39,10 @@ void Session::Execute() {
   }
 
   // Remove the data from the stream on scope exit.
-  utils::OnScopeExit shift_data([&, ret] { input_stream_->Shift(ret.stream_size); });
+  auto const shift_data = utils::OnScopeExit{[&, ret] {
+    input_stream_->Shift(ret.stream_size);
+    input_stream_->ShrinkBuffer(kBufferRetainLimit);
+  }};
```

Comment:
I didn't go to nitty-gritty details here but ret.stream_size can be bigger than 4MB when getting data via stream?

Reply:
This is OnScopeExit straight after the Shift that has remove exactly `ret.stream_size` from the start of the buffer. Shift would AFAICT should have reset stream back to have 0. `ShrinkBuffer` will be to the max of kBufferRetainLimit and have, likely kBufferRetainLimit, and adjust accordingly. 
I have gone back and changed the implementation in order to only copy `have_` amount upon resize.